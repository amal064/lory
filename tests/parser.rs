use insta::assert_debug_snapshot;
use lory::{lexer, parser};

macro_rules! test_expression_from_file {
    ($file:expr) => {{
        let tokens = lexer::tokenize(include_str!($file));
        let mut parser = parser::Parser::new(tokens);
        let ast = parser.parse_expression();
        assert_debug_snapshot!(ast);
    }};
}

#[test]
fn expression() {
    test_expression_from_file!("parser/expression.lox");
}
