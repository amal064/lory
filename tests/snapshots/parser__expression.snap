---
source: tests/parser.rs
expression: ast
---
Ok(
    BinaryOp {
        left: BinaryOp {
            left: Literal(
                Number(
                    5.0,
                ),
            ),
            op: Sub,
            right: BinaryOp {
                left: Literal(
                    Number(
                        3.0,
                    ),
                ),
                op: Sub,
                right: Literal(
                    Number(
                        1.0,
                    ),
                ),
            },
        },
        op: Add,
        right: UnaryOp {
            op: Neg,
            expr: Literal(
                Number(
                    1.0,
                ),
            ),
        },
    },
)
